pkg/cluster/kubeadm-join.go

```diff
@@ -0,0 +1,102 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/kind/pkg/cluster/config"
+	"sigs.k8s.io/kind/pkg/cluster/kubeadm"
+)
+
+// KubeadmJoinAction implements action for joining nodes
+// to a Kubernetes cluster.
+type KubeadmJoinAction struct{}
+
+func init() {
+	RegisterAction("join", NewKubeadmJoinAction)
+}
+
+// NewKubeadmJoinAction returns a new KubeadmJoinAction
+func NewKubeadmJoinAction() Action {
+	return &KubeadmJoinAction{}
+}
+
+// Tasks returns the list of action tasks
+func (b *KubeadmJoinAction) Tasks() []Task {
+	return []Task{
+		// TODO(fabrizio pandini): add Run kubeadm join --experimental-master
+		//      on SecondaryControlPlaneNodes
+		{
+			// Run kubeadm join on the WorkeNodes
+			Description: "Joining worker node to Kubernetes ☸",
+			TargetNodes: SelectWorkerNodes,
+			Run:         runKubeadmJoin,
+		},
+	}
+}
+
+// runKubeadmJoin executes kubadm join
+func runKubeadmJoin(ec *execContext, configNode *config.Node) error {
+	// before running join, it should be retrived
+
+	// gets the node where
+	// TODO(fabrizio pandini): when external load-balancer will be
+	//      implemented this should be modified accordingly
+	controlPlaneHandle, ok := ec.NodeFor(ec.config.BootStrapControlPlane())
+	if !ok {
+		return fmt.Errorf("unable to get the handle for operating on node: %s", ec.config.BootStrapControlPlane().Name)
+	}
+
+	// gets the IP of the bootstrap master node
+	controlPlaneIP, err := controlPlaneHandle.IP()
+	if err != nil {
+		return errors.Wrap(err, "failed to get IP for node")
+	}
+
+	// get the target node for this task
+	node, ok := ec.NodeFor(configNode)
+	if !ok {
+		return fmt.Errorf("unable to get the handle for operating on node: %s", configNode.Name)
+	}
+
+	// TODO(fabrizio pandini): might be we want to run pre-kubeadm hooks on workers too
+
+	// run kubeadm
+	if err := node.Command(
+		"kubeadm", "join",
+		// the control plane address uses the docker ip and a well know APIServerPort that
+		// are accessible only inside the docker network
+		fmt.Sprintf("%s:%d", controlPlaneIP, kubeadm.APIServerPort),
+		// uses a well known token and skipping ca certification for automating TLS bootstrap process
+		"--token", kubeadm.Token,
+		"--discovery-token-unsafe-skip-ca-verification",
+		// preflight errors are expected, in particular for swap being enabled
+		// TODO(bentheelder): limit the set of acceptable errors
```

Comment:
is this code (and code like this) copied from another location but is not removed from that original location in the diff?

Reply:
This code is intentionally copied here because it is the same problem/TODO already defined for init and they should be addressed with the same solution

But I agree, the diff for context.go looks ugly and doesn't help...